text/3318-field-projection.md

```diff
@@ -0,0 +1,400 @@
+fn get_field<F: Field<Base = Problem>>(problem: &Problem) -> &F::Type {
+    let ptr: *const Problem = problem;
+    // SAFETY: `F` implements the `Field` trait and thus we find `F::Type` at `F::OFFSET` inside
+    // of `ptr` that was derived from a reference.
+    unsafe { &*ptr.cast::<u8>().add(F::OFFSET).cast::<F::Type>() }
+}
+```
+There are a lot more powerful things that one can do using this type. For example field projections can be expressed safely. If we are often working with memory that has to be accessed volatile, then we might write the following wrapper type:
+```rust
+/// A pointer to memory that enforces volatile access.
+pub struct VolatileMem<T> {
+    ptr: NonNull<T>,
+}
+
+impl<T: Copy> VolatileMem<T> {
+    pub fn get(&self) -> T {
+        // SAFETY: `ptr` is always valid for volatile reads.
+        unsafe { ptr::read_volatile(self.ptr.as_ptr()) }
+    }
+
+    pub fn put(&mut self, val: T) {
+        // SAFETY: `ptr` is always valid for volatile writes.
+        unsafe { ptr::write_volatile(self.ptr.as_mut_ptr()) }
+    }
+}
+```
+Now consider the following struct that we would like to put into our `VolatileMem<T>`:
+```rust
+#[repr(C)]
+pub struct Config {
+    mode: u8,
+    reserved: [u8; 128],
+}
+```
+If we want to write a new config, then we always have to write the whole struct, including the `reserved` field that is comparatively big. We can avoid this by providing a field projection:
+```rust
+impl<T> VolatileMem<T> {
+    pub fn map<F: Field<Base = T>>(self) -> VolatileMem<F::Type> {
+        Self {
+            // SAFETY: `F` implements the `Field` trait and thus we find `F::Type` at `F::OFFSET`
+            // inside of `ptr` that is always valid.
+            ptr: unsafe {
+                NonNull::new_unchecked(
+                    self.ptr.as_ptr().cast::<u8>().add(F::OFFSET).cast::<F::Type>(),
+                )
+            },
+        }
+    }
+}
+```
+Now in the scenario from above we can do:
+```rust
+let mut config: VolatileMem<Config> = ...;
+config.put(Config::default());
+let mut mode: VolatileMem<u8> = config.map::<field_of!(Config, mode)>();
+mode.put(1);
+```
+And we will not have to always overwrite `reserved` with the same data.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+For every field of every non-packed struct and tuple, the compiler creates a unique, unnameable type that represent that field. These generated types will:
+- have meaningful names in error messages (e.g. `Struct::field`, `(i32, u32)::0`), 
+- implement the `Field` with accurate associated types and constants.
+
+The `Field` trait will reside in `core::marker` and is:
+```rust
+/// A type representing a field on a struct.
+pub trait Field {
+    /// The type (struct or tuple) containing this field.
+    type Base;
+    /// The type of this field.
+    type Type;
+    /// The offset of this field from the beginning of the `Base` type in bytes.
+    const OFFSET: usize;
```

Comment:
How is this defined when the field's alignment is not known at compile-time? (Like `struct Foo { a: i32, b: dyn Debug }`)

Reply:
What does the alignment have to do with the offset? IIRC in your example `a` should have offset 0 and `b` should have offset 4.

Reply:
For example, if the `dyn Debug` stored in `b` is really `u64`, it would be stored starting at byte 8, not 4. But you can only know this at runtime, from the vtable.

Reply:
Ah I see, I think we can just copy `offset_of!` here.

Reply:
I think we should still generate a type for the field, just not implement the `Field` trait for it.